pkg/ddl/ddl_running_jobs.go

```diff
@@ -19,122 +19,158 @@
+	// zero-length map exists in table-level.
+	schemas map[string]map[string]struct{}
 }
 
 func newRunningJobs() *runningJobs {
 	return &runningJobs{
-		processingIDs:    make(map[int64]struct{}),
-		unfinishedSchema: make(map[string]map[string]struct{}),
-		unfinishedIDs:    make(map[int64]struct{}),
+		ids:          make(map[int64]struct{}),
+		idsStrGetter: func() string { return "" },
+		schemas:      make(map[string]map[string]struct{}),
 	}
 }
 
-func (j *runningJobs) clear() {
-	j.Lock()
-	defer j.Unlock()
-	j.unfinishedIDs = make(map[int64]struct{})
-	j.unfinishedSchema = make(map[string]map[string]struct{})
-}
+// add should only add the argument that passed the last checkRunnable.
+func (j *runningJobs) add(jobID int64, involves []model.InvolvingSchemaInfo) {
+	j.mu.Lock()
+	defer j.mu.Unlock()
 
-func (j *runningJobs) add(jobID int64, involvedSchemaInfos []model.InvolvingSchemaInfo) {
-	j.Lock()
-	defer j.Unlock()
-	j.processingIDs[jobID] = struct{}{}
-	j.updateInternalRunningJobIDs()
+	j.ids[jobID] = struct{}{}
+	j.updateIDsStrGetter()
 
-	if _, ok := j.unfinishedIDs[jobID]; ok {
-		// Already exists, no need to add it again.
-		return
-	}
-	j.unfinishedIDs[jobID] = struct{}{}
-	for _, info := range involvedSchemaInfos {
-		if _, ok := j.unfinishedSchema[info.Database]; !ok {
-			j.unfinishedSchema[info.Database] = make(map[string]struct{})
+	for _, info := range involves {
+		if info.Database == model.InvolvingNone {
+			// should not happen
+			if intest.InTest {
+				if info.Table != model.InvolvingNone {
+					panic(fmt.Sprintf(
+						"job %d is invalid, involved table name is not empty: %s",
+						jobID, info.Table,
+					))
+				}
+			}
+			continue
+		}
+
+		if _, ok := j.schemas[info.Database]; !ok {
+			j.schemas[info.Database] = make(map[string]struct{})
 		}
-		j.unfinishedSchema[info.Database][info.Table] = struct{}{}
+		j.schemas[info.Database][info.Table] = struct{}{}
 	}
 }
 
-func (j *runningJobs) remove(jobID int64, involvedSchemaInfos []model.InvolvingSchemaInfo) {
-	j.Lock()
-	defer j.Unlock()
-	delete(j.processingIDs, jobID)
-	j.updateInternalRunningJobIDs()
+// remove can be concurrently called with add and checkRunnable.
+func (j *runningJobs) remove(jobID int64, involves []model.InvolvingSchemaInfo) {
+	j.mu.Lock()
+	defer j.mu.Unlock()
 
-	delete(j.unfinishedIDs, jobID)
-	for _, info := range involvedSchemaInfos {
-		if db, ok := j.unfinishedSchema[info.Database]; ok {
+	if intest.InTest {
+		if _, ok := j.ids[jobID]; !ok {
+			panic(fmt.Sprintf("job %d is not running", jobID))
+		}
+	}
+	delete(j.ids, jobID)
+	j.updateIDsStrGetter()
+
+	for _, info := range involves {
+		if db, ok := j.schemas[info.Database]; ok {
 			delete(db, info.Table)
 		}
-		if len(j.unfinishedSchema[info.Database]) == 0 {
-			delete(j.unfinishedSchema, info.Database)
+		if len(j.schemas[info.Database]) == 0 {
+			delete(j.schemas, info.Database)
 		}
 	}
 }
 
+func (j *runningJobs) updateIDsStrGetter() {
+	var (
+		once   sync.Once
```

Comment:
`j.idsStrGetter()` has been protected by `j.RLock()`, it is still necessary to use `sync.Once`? I mean why not:

```
if idStr == nil {
    // construct idstr
}
return idsStr
```

Reply:
RLock will let multiple lock holders write to idsStr, which is a data race